File: java/src/test/java/org/sammancoaching/parameterizedtests/concrete/OrderStateTest.java
package org.sammancoaching.parameterizedtests.concrete;

import static org.sammancoaching.parameterizedtests.Status.PAID;
import static org.sammancoaching.parameterizedtests.Status.PAYMENT_EXPECTED;
import static org.sammancoaching.parameterizedtests.Status.PREPARING;
import static org.sammancoaching.parameterizedtests.Status.READY;
import static org.sammancoaching.parameterizedtests.Status.TAKEN;
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

import org.sammancoaching.parameterizedtests.Order;
import org.sammancoaching.parameterizedtests.Status;

class OrderStateTest {

	@Test
	void orderCanBeUpdatedAndCanceledIfNotPaidYet() {
		Order order = existingOrderWithStatus(PAYMENT_EXPECTED);
		assertThat(order.isUpdateAllowed()).isTrue();
		assertThat(order.isCancelAllowed()).isTrue();
	}

	@Test
	void orderCannotBeUpdatedOrCanceledIfPaid() {
		Order order = existingOrderWithStatus(PAID);
		assertThat(order.isUpdateAllowed()).isFalse();
		assertThat(order.isCancelAllowed()).isFalse();
	}

	@Test
	void orderCannotBeUpdatedOrCanceledIfPreparing() {
		Order order = existingOrderWithStatus(PREPARING);
		assertThat(order.isUpdateAllowed()).isFalse();
		assertThat(order.isCancelAllowed()).isFalse();
	}

	@Test
	void orderCannotBeUpdatedOrCanceledIfReady() {
		Order order = existingOrderWithStatus(READY);
		assertThat(order.isUpdateAllowed()).isFalse();
		assertThat(order.isCancelAllowed()).isFalse();
	}

	@Test
	void orderCannotBeUpdatedOrCanceledIfTaken() {
		Order order = existingOrderWithStatus(TAKEN);
		assertThat(order.isUpdateAllowed()).isFalse();
		assertThat(order.isCancelAllowed()).isFalse();
	}

	private static Order existingOrderWithStatus(Status status) {
		Order order = new Order(1L);
		order.setStatus(status);
		return order;
	}
}

File: typescript/tests/order-state.test.ts
import { describe, expect, test } from "bun:test";
import { Order } from "@/order";
import { Status } from "@/status";

function existingOrderWithStatus(status: Status): Order {
	const order = new Order(1);
	order.setStatus(status);
	return order;
}

describe("OrderStateTest", () => {
	test("order can be updated and canceled if not paid yet", () => {
		const order = existingOrderWithStatus(Status.PAYMENT_EXPECTED);
		expect(order.isUpdateAllowed()).toBe(true);
		expect(order.isCancelAllowed()).toBe(true);
	});

	test("order cannot be updated or canceled if paid", () => {
		const order = existingOrderWithStatus(Status.PAID);
		expect(order.isUpdateAllowed()).toBe(false);
		expect(order.isCancelAllowed()).toBe(false);
	});

	test("order cannot be updated or canceled if preparing", () => {
		const order = existingOrderWithStatus(Status.PREPARING);
		expect(order.isUpdateAllowed()).toBe(false);
		expect(order.isCancelAllowed()).toBe(false);
	});

	test("order cannot be updated or canceled if ready", () => {
		const order = existingOrderWithStatus(Status.READY);
		expect(order.isUpdateAllowed()).toBe(false);
		expect(order.isCancelAllowed()).toBe(false);
	});

	test("order cannot be updated or canceled if taken", () => {
		const order = existingOrderWithStatus(Status.TAKEN);
		expect(order.isUpdateAllowed()).toBe(false);
		expect(order.isCancelAllowed()).toBe(false);
	});
});
